avm/src/vm/event.rs

```diff
@@ -192,6 +195,80 @@ impl HandlerFragment {
     };
   }
 
+  #[inline(always)]
+  async fn run_cpu(
+    &mut self,
+    mut hand_mem: Arc<HandlerMemory>,
+    instrs: &Vec<Instruction>
+  ) -> Arc<HandlerMemory> {
+    instrs.iter().for_each(|i| {
+      if let OpcodeFn::Cpu(func) = i.opcode.fun {
+        //eprintln!("{} {:?}", i.opcode._name, i.args);
+        if let Some(event) = func(i.args.as_slice(), &mut hand_mem) {
+          let event_tx = EVENT_TX.get().unwrap();
+          if let Err(_) = event_tx.send(event) {
+            eprintln!("Event transmission error");
+            std::process::exit(1);
+          }
+        };
+      } else {
+        eprintln!("expected another CPU instruction");
+        std::process::exit(1);
+      }
+    });
+    hand_mem
+  }
+
+  #[inline(always)]
+  async fn run_unpred_cpu(
+    &mut self,
+    hand_mem: Arc<HandlerMemory>,
+    instrs: &Vec<Instruction>
+  ) -> Arc<HandlerMemory> {
+    // These instructions are always in groups by themselves
+    let op = &instrs[0];
+    if let OpcodeFn::UnpredCpu(func) = op.opcode.fun {
```

Comment:
so an `unpred_cpu` can never emit an event?

Reply:
`emit` is a `cpu` opcode so it gets handled by `run_cpu`

Reply:
I've honestly never liked how we handled event emission in the opcodes. There's really only one opcode that ever emits an event, but *all* of the `cpu!` opcodes have to deal with that. `unpred_cpu` opcodes by themselves are unpredictable because they internally run another handler (and we want all of their mutations to stick around). One or more of the opcodes within that handler can `emit` just fine?

Reply:
That makes sense. I agree with this:
```
I've honestly never liked how we handled event emission in the opcodes. There's really only one opcode that ever emits an event, but all of the cpu! opcodes have to deal with that.
```